src/modules/m_sqlauth.cpp

```diff
@@ -34,22 +34,41 @@ class AuthQuery : public SQLQuery
 	const std::string uid;
 	LocalIntExt& pendingExt;
 	bool verbose;
-	AuthQuery(Module* me, const std::string& u, LocalIntExt& e, bool v)
-		: SQLQuery(me), uid(u), pendingExt(e), verbose(v)
+	const std::string kdf;
+	AuthQuery(Module* me, const std::string& u, LocalIntExt& e, bool v, const std::string& kd)
+		: SQLQuery(me), uid(u), pendingExt(e), verbose(v), kdf(kd)
 	{
 	}
 
 	void OnResult(SQLResult& res) CXX11_OVERRIDE
 	{
-		User* user = ServerInstance->FindNick(uid);
+		LocalUser* user = dynamic_cast<LocalUser*>(ServerInstance->FindNick(uid));
 		if (!user)
 			return;
+
 		if (res.Rows())
 		{
+			if (!kdf.empty())
+			{
+				HashProvider* hashprov = ServerInstance->Modules->FindDataService<HashProvider>("hash/" + kdf);
+				if (!hashprov)
+					goto authfailure;
```

Comment:
I don't like this goto, and it also jumps to log the wrong thing ("SQL query returned no matches")